# Graded assignment 1

Name: Borná Djavdan  
SNR: 2007904


# Packages

We need the package `tidyverse`

In [14]:
library(tidyverse) 

# Assignment 1

First, we generate some random data:

In [15]:
df1 <- as.data.frame(matrix(runif(20*2), ncol=20) )
head(df1)

V1          V2        V3        V4        V5        V6         V7        
1 0.914802879 0.9270443 0.3152392 0.3433697 0.1775565 0.07647695 0.63572248
2 0.007817592 0.3986331 0.5687628 0.8913555 0.9188985 0.46784472 0.08149938
  V8        V9        V10       V11         V12       V13       V14      
1 0.5732052 0.2724442 0.9429531 0.003987219 0.2890462 0.4831759 0.1742860
2 0.4295361 0.8978777 0.6560009 0.255052113 0.9712480 0.2086056 0.2108421
  V15       V16       V17       V18       V19       V20      
1 0.2931146 0.1706397 0.6640952 0.4564880 0.7054454 0.2636715
2 0.3054049 0.2700215 0.4649118 0.4923848 0.7549877 0.5171978

_df1_ is a wide dataframe. The goal of the assignment is to change _df1_ into a long dataframe _df2_. And show the first rows of the dataframe using `head(df2)`

The dataframe _df2_ should look like this:

| Key | Value      |
|-----|------------|
| V1  | 0.09413864 |
| V1  | 0.69782526 |
| V2  | 0.23460454 |
| V2  | 0.51520514 |
| V3  | 0.63181844 |
| V3  | 0.22807211 |

In [16]:
df2 <- df1 %>%
  gather(Key, Value, V1:V20)
head(df2)

Key Value      
1 V1  0.914802879
2 V1  0.007817592
3 V2  0.927044331
4 V2  0.398633063
5 V3  0.315239163
6 V3  0.568762786

## Assignment 2

We will generate some data:

In [17]:
df3 <- as.data.frame(read.table(header=TRUE, text='
      id        V_1       V_2                
          1  42.467573  -87.810047     
          2  42.049148  -88.273029
          3  39.110539  -90.324080
                       '))

head(df3)

id V_1      V_2      
1 1  42.46757 -87.81005
2 2  42.04915 -88.27303
3 3  39.11054 -90.32408

In [18]:
df4 <- as.data.frame(read.table(header=TRUE, text='
     V_3            V_4         id                 
      one         1h            1
      two         1h            1
      three       8h            2
      four        1h            4
'))

head(df4)

V_3   V_4 id
1 one   1h  1 
2 two   1h  1 
3 three 8h  2 
4 four  1h  4


You will have to merge these files, returning all rows and all columns from both _df3_ and _df5_. Where there are not matching values, return _NA_ for the missings. 

This should be the end result:

| id | V_1      | V_2       | V_3   | V_4 |
|----|----------|-----------|-------|-----|
| 1  | 42.46757 | -87.81005 | one   | 1h  |
| 1  | 42.46757 | -87.81005 | two   | 1h  |
| 2  | 42.04915 | -88.27303 | three | 8h  |
| 3  | 39.11054 | -90.32408 | NA    | NA  |
| 4  | NA       | NA        | four  | 1h  |



I like `full_join` better in this case: 

In [19]:
df5 <- full_join(df3, df4, by='id')
df5

id V_1      V_2       V_3   V_4
1 1  42.46757 -87.81005 one   1h 
2 1  42.46757 -87.81005 two   1h 
3 2  42.04915 -88.27303 three 8h 
4 3  39.11054 -90.32408 NA    NA 
5 4        NA        NA four  1h

You can also do it by using `merge`:

In [20]:
df5 <- merge(df3, df4, all="true")
df5

id V_1      V_2       V_3   V_4
1 1  42.46757 -87.81005 one   1h 
2 1  42.46757 -87.81005 two   1h 
3 2  42.04915 -88.27303 three 8h 
4 3  39.11054 -90.32408 NA    NA 
5 4        NA        NA four  1h

End of notebook